File: src/main/BestTimeToBuyAndSellStock.java
/**
 * 买卖股票的最佳时机：https://leetcode-cn.com/problems/best-time-to-buy-and-sell-stock/
 */
public class BestTimeToBuyAndSellStock {

    public static int maxProfit(int[] prices) {
        int minPrice = Integer.MAX_VALUE;
        int maxProfit = 0;
        for (int i = 0; i < prices.length; i++) {
            minPrice = Math.min(minPrice, prices[i]);
            maxProfit = Math.max(maxProfit, prices[i] - minPrice);
        }
        return maxProfit;
    }

    public static void main(String[] args) {
        System.out.println(maxProfit(new int[] {7,1,5,3,6,4}));
        System.out.println(maxProfit(new int[] {7,6,4,3,1}));
        System.out.println(maxProfit(new int[] {2,4,1}));
    }

}

File: c/best-time-to-buy-and-sell-stock.c
/**
 * 121. 买卖股票的最佳时机：https://leetcode.cn/problems/best-time-to-buy-and-sell-stock
 */

#define MIN(a,b) (((a)<(b))?(a):(b))
#define MAX(a,b) (((a)>(b))?(a):(b))

int maxProfit(int *prices, int n) {
    int ans = 0;
    int min = prices[0];
    for (int i = 1; i < n; i++) {
        min = MIN(min, prices[i]);
        ans = MAX(ans, prices[i] - min);
    }
    return ans;
}
